stdlib/args.ysh

```diff
@@ -24,140 +24,159 @@
-
-  # TODO:
-  # - need :: because vars start with _
-  # - It would be nice if _this was a magic param?
-  #   - it would be nice if 'flag' was a "method" on something?
-  #   - that's what it's doing, it's mutating "self"
-  #   - _this is the thing we're currently creating
-
-  _ _this.flags->append(result)
-
-  var this = getvar('_this', scope='dynamic')
-  _ this.flags->append(result)
 }
 
-proc __arg (name, desc ; type ; default Any = null) {
-  # Type is always string?  Or support conversion?
-  # I guess there's no reason not to
-  #
-  # Int or Float
-
-  echo 'arg'
+proc flag (short, long; type='bool', default=null) {
+  ## Declare a flag within an `Args`.
+  ##
+  ## Examples:
+  ##
+  ##   Args :spec {
+  ##     flag -v --verbose
+  ##     flag -n --count ('int', default=1)
+  ##     flag -f --file ('str')
+  ##   }
+
+  # TODO: help, make type take in a type object (ie. Bool, Int, Str)
+  json write --pretty=F ({ tag: 'flag', node: { short, long, type, default, help: null } })
 }
 
-# TODO: create a builtin?
-func getProc(x) {
-  return (x)
+proc arg (name) {
+  ## Declare a positional argument within an `Args`.
+  ##
+  ## Examples:
+  ##
+  ##   Args :spec {
+  ##     arg file
+  ##   }
+
+  # TODO: type, default and help
+  json write --pretty=F ({ tag: 'arg', node: { name, type: 'str', default: null, help: null } })
 }
-# TODO: builtin?
-# Is this different than shvar?  How?
-# Is it a global or a local?
 
-# Maybe we can rename shvar IFS=: (foo={}) { }
-#
-# - shvar does LocalOnly
-# - but 'proc arg' will push a new stack frame?
-#   - so we need something like setref?
-#
-
-proc bind-vars (; ; ...bindings; block) {
-  echo
+proc rest (name) {
+  ## Take the remaining positional arguments within an `Args`.
+  ##
+  ## Examples:
+  ##
+  ##   Args :grep {
+  ##     arg query
+  ##     rest files
+  ##   }
+
+  json write --pretty=F ({ tag: 'rest', node: name })
 }
 
-hay define Documented
-
-Documented {  # Do we inherit the name of the proc?  Or $path/$procName?
-
-  # Do we want a longer way of documenting APIs?
-  # Document every argument?
-  # I like Go's docs
-  desc = '''
-  These are my docs
-  '''
-
-  proc p {
-    echo
+func __args_getFlagName(flag) {
+  if (flag.long) {
+    return (flag.long[2:])
+  } elif (flag.short) {
+    return (flag.short[1:])
   }
 
+  error ('No flag.long or flag.short')
 }
 
-proc Args(spec Ref ; ; ; block) {
-
-  var more_procs = {
-    flag: getProc('__flag'),
-    arg: getProc('__arg'),
-  }
-
-  var this = {}
+func parseArgs(spec, argv) {
+  ## Given a spec created by `Args`. Parse an array of strings `argv` per that
+  ## spec.
+  ##
+  ## See `Args` for examples of use.
 
-  # _this is part of this frame -- it's a duplicate that callers can look up I
-  # guess
-  # Well then all you need is getvar() ?  You don't even need bind-vars?
-  bind-vars (_this=this) {
-    # So now 'arg' and 'flag' can see _this?
+  var i = 0
+  var positionalPos = 0
```

Comment:
On Zulip we mentioned `snake_case` for locals, and then `camelCase()` for functions

It is a little funny that it feels like `func` wants `namesLikeJS`, while proc wants `names_like_shell`, but that is probably too fine a distinction